src/Controllers/ApiController.php

```diff
@@ -27,4 +39,114 @@ public function index()
             ->withHeader('content-type', 'application/json')
             ->withContent(json_encode(['error' => 'Not implemented']));
     }
+
+
+    /**
+     * @return User
+     *
+     * Warning: I am no php coder. This code isnt intended for eternity but making our lives @36c3 a bit easier.
+     * Code is mostly harvested from /includes/controller/shifts_controller.php/shifts_json_export_controller()
+     */
+    protected function doApiAuth($key)
+    {
+        $user = auth()->apiUser('key');
+        if (
+            !preg_match('/^[\da-f]{32}$/', $key)
+            || !$user
+        ) {
+            throw new HttpForbidden('{"error":"Missing or invalid key"}', ['content-type' => 'application/json']);
+        }
+
+        if (!auth()->can('view_api')) {
+            throw new HttpForbidden('{"error":"Not allowed"}', ['content-type' => 'application/json']);
+        }
+
+        return $user;
+    }
+
+    /**
+     * @Route(“/api/shifts/my”)
+     */
+    public function getMyShifts(Request $request)
+    {
+        $key = $request->input('key');
+        $user = $this->doApiAuth($key);
+
+        $shifts = Shifts_by_user($user->id);
+
+        return $this->response
+            ->setStatusCode(200)
+            ->withHeader('content-type', 'application/json')
+            ->withContent(json_encode($shifts));
+    }
+
+    /**
+     * @Route(“/api/shifts/my”)
+     * doesnt work? dunno why result is empty.
+     */
+    public function getShiftsByAngelType(Request $request)
+    {
+        $key = $request->input('key');
+        $user = $this->doApiAuth($key);
+
+        // Shifts_by_angeltype uses only the id, required a AngelType though.
+        $angelTypeId = (int)$request->getAttribute('angeltypeid');
+        $angelType = AngelType($angelTypeId);
+        $shifts = Shifts_by_angeltype($angelType);
+
+        return $this->response
+            ->setStatusCode(200)
+            ->withHeader('content-type', 'application/json')
+            ->withContent(json_encode($shifts));
+    }
+
+    /**
+     * @Route(“/api/shifts/free”)
+     */
+    public function getShiftsFree(Request $request)
+    {
+        $key = $request->input('key');
+        $user = $this->doApiAuth($key);
+        // Shifts_by_angeltype uses only the id, required a AngelType though.
+        $start = (int)$request->getAttribute('start');
```

Comment:
Using timestamps might be the easiest thing (as the legacy database model uses that) but using a datetime is more readable and thus should be imho preferred

Reply:
I'm still uncertain about this. While it's unreadable for humans, it's easy to use in any kind of code and you instantly recognize it as a timestamp.
String representations of datetimes are often annoying due to a quadrillion different format variations even though human readable (but not necessary unambiguously)

Reply:
As thats only an API i think we could just define that a datetime has to be `DateTimeInterface::RFC3339` formatted and thus look like `'Y-m-d\TH:i:sP'` which should solve the interpretation problem

Reply:
Why not both ;)